src/common/providers/identity.ts

```diff
@@ -109,3 +375,598 @@ export function userRecordConstructor(wireData: Object): UserRecord {
+    }
+  if (profile) {
+    if (decodedJWT.sign_in_method === 'github.com') {
+      username = profile.login;
+    }
+    if (decodedJWT.sign_in_method === 'twitter.com') {
+      username = profile.screen_name;
+    }
+  }
+
+  return {
+    providerId:
+      decodedJWT.sign_in_method === 'emailLink'
+        ? 'password'
+        : decodedJWT.sign_in_method,
+    profile,
+    username,
+    isNewUser: decodedJWT.event_type === 'beforeCreate' ? true : false,
+  };
+}
+
+/** @internal */
+function parseAuthCredential(decodedJWT: DecodedJwt, time: number): Credential {
+  if (
+    !decodedJWT.sign_in_attributes &&
+    !decodedJWT.oauth_id_token &&
+    !decodedJWT.oauth_access_token &&
+    !decodedJWT.oauth_refresh_token
+  ) {
+    return null;
+  }
+  return {
+    claims: decodedJWT.sign_in_attributes,
+    idToken: decodedJWT.oauth_id_token,
+    accessToken: decodedJWT.oauth_access_token,
+    refreshToken: decodedJWT.oauth_refresh_token,
+    expirationTime: decodedJWT.oauth_expires_in
+      ? new Date(time + decodedJWT.oauth_expires_in * 1000).toUTCString()
+      : undefined,
+    secret: decodedJWT.oauth_token_secret,
+    providerId:
+      decodedJWT.sign_in_method === 'emailLink'
+        ? 'password'
+        : decodedJWT.sign_in_method,
+    signInMethod: decodedJWT.sign_in_method,
+  };
+}
+
+/**
+ * @internal
+ * Parses the decoded jwt into a valid AuthEventContext for use in the handler
+ */
+export function parseAuthEventContext(
+  decodedJWT: DecodedJwt,
+  projectId: string,
+  time: number = new Date().getTime()
+): AuthEventContext {
+  const eventType =
+    (EVENT_MAPPING[decodedJWT.event_type] || decodedJWT.event_type) +
+    (decodedJWT.sign_in_method ? `:${decodedJWT.sign_in_method}` : '');
+
+  return {
+    locale: decodedJWT.locale,
+    ipAddress: decodedJWT.ip_address,
+    userAgent: decodedJWT.user_agent,
+    eventId: decodedJWT.event_id,
+    eventType,
+    authType: !!decodedJWT.user_record ? 'USER' : 'UNAUTHENTICATED',
+    resource: {
+      // TODO(colerogers): figure out the correct service
+      service: 'identitytoolkit.googleapis.com',
+      name: !!decodedJWT.tenant_id
+        ? `projects/${projectId}/tenants/${decodedJWT.tenant_id}`
+        : `projects/${projectId}`,
+    },
+    timestamp: new Date(decodedJWT.iat * 1000).toUTCString(),
+    additionalUserInfo: parseAdditionalUserInfo(decodedJWT),
+    credential: parseAuthCredential(decodedJWT, time),
+    params: {},
+  };
+}
+
+/**
+ * @internal
+ * Checks the handler response for invalid customClaims & sessionClaims objects
+ */
+export function validateAuthResponse(
+  eventType: string,
+  authRequest?: BeforeCreateResponse | BeforeSignInResponse
+) {
+  const nonAllowListedClaims = [
+    'acr',
+    'amr',
+    'at_hash',
+    'aud',
+    'auth_time',
+    'azp',
+    'cnf',
+    'c_hash',
+    'exp',
+    'iat',
+    'iss',
+    'jti',
+    'nbf',
+    'nonce',
+    'firebase',
+  ];
+  const claimsMaxPayloadSize = 1000;
```

Comment:
I wonder if these should be placed as global case and made UPPER_SNAKE_CASE

Reply:
Good point